src/Microsoft.AspNetCore.TestHost/HttpContextBuilder.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting.Server;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using static Microsoft.AspNetCore.Hosting.Internal.HostingApplication;
+
+namespace Microsoft.AspNetCore.TestHost
+{
+    internal class HttpContextBuilder
+    {
+        private readonly IHttpApplication<Context> _application;
+        private readonly HttpContext _httpContext;
+        
+        private TaskCompletionSource<HttpContext> _responseTcs = new TaskCompletionSource<HttpContext>(TaskCreationOptions.RunContinuationsAsynchronously);
+        private ResponseStream _responseStream;
+        private ResponseFeature _responseFeature = new ResponseFeature();
+        private CancellationTokenSource _requestAbortedSource = new CancellationTokenSource();
+        private bool _pipelineFinished;
+        private Context _testContext;
+
+        internal HttpContextBuilder(IHttpApplication<Context> application)
+        {
+            _application = application ?? throw new ArgumentNullException(nameof(application));
+            _httpContext = new DefaultHttpContext();
+
+            var request = _httpContext.Request;
+            request.Protocol = "HTTP/1.1";
+            request.Method = HttpMethods.Get;
+
+            _httpContext.Features.Set<IHttpResponseFeature>(_responseFeature);
+            var requestLifetimeFeature = new HttpRequestLifetimeFeature();
+            requestLifetimeFeature.RequestAborted = _requestAbortedSource.Token;
+            _httpContext.Features.Set<IHttpRequestLifetimeFeature>(requestLifetimeFeature);
+            
+            _responseStream = new ResponseStream(ReturnResponseMessageAsync, AbortRequest);
+            _responseFeature.Body = _responseStream;
+        }
+
+        internal void Configure(Action<HttpContext> configureContext)
+        {
+            if (configureContext == null)
+            {
+                throw new ArgumentNullException(nameof(configureContext));
+            }
+
+            configureContext(_httpContext);
```

Comment:
`=> configureContext?.Invoke(_httpContext) ?? throw new ArgumentNullException(nameof(configureContext));`?

Reply:
Yuck

Reply:
I second that yuck